torchaudio/prototype/pipelines/_vggish/_vggish_impl.py

```diff
@@ -0,0 +1,262 @@
+            0.0, nyquist_hertz, num_spectrogram_bins
+        )
+        spectrogram_bins_mel = self._hertz_to_mel(spectrogram_bins_hertz)
+        # The i'th mel band (starting from i=1) has center frequency
+        # band_edges_mel[i], lower edge band_edges_mel[i-1], and higher edge
+        # band_edges_mel[i+1].  Thus, we need num_mel_bins + 2 values in
+        # the band_edges_mel arrays.
+        band_edges_mel = torch.linspace(
+            self._hertz_to_mel(torch.tensor(lower_edge_hertz)),
+            self._hertz_to_mel(torch.tensor(upper_edge_hertz)),
+            num_mel_bins + 2,
+        )
+        # Matrix to post-multiply feature arrays whose rows are num_spectrogram_bins
+        # of spectrogram values.
+        mel_weights_matrix = torch.empty((num_spectrogram_bins, num_mel_bins))
+        for i in range(num_mel_bins):
+            lower_edge_mel, center_mel, upper_edge_mel = band_edges_mel[i : i + 3]
+            # Calculate lower and upper slopes for every spectrogram bin.
+            # Line segments are linear in the *mel* domain, not hertz.
+            lower_slope = (spectrogram_bins_mel - lower_edge_mel) / (
+                center_mel - lower_edge_mel
+            )
+            upper_slope = (upper_edge_mel - spectrogram_bins_mel) / (
+                upper_edge_mel - center_mel
+            )
+            # .. then intersect them with each other and zero.
+            mel_weights_matrix[:, i] = torch.maximum(
+                torch.tensor(0.0), torch.minimum(lower_slope, upper_slope)
+            )
+        # HTK excludes the spectrogram DC bin; make sure it always gets a zero
+        # coefficient.
+        mel_weights_matrix[0, :] = 0.0
+        return mel_weights_matrix
+
+    def _log_mel_spectrogram(
+        self,
+        data,
+        audio_sample_rate=8000,
+        log_offset=0.0,
+        window_length_secs=0.025,
+        hop_length_secs=0.010,
+        **kwargs,
+    ):
+        window_length_samples = int(round(audio_sample_rate * window_length_secs))
+        hop_length_samples = int(round(audio_sample_rate * hop_length_secs))
+        fft_length = 2 ** int(
+            math.ceil(math.log(window_length_samples) / math.log(2.0))
+        )
+        spectrogram = self._stft_magnitude(
+            data,
+            fft_length=fft_length,
+            hop_length=hop_length_samples,
+            window_length=window_length_samples,
+        )
+        mel_spectrogram = torch.matmul(
+            spectrogram,
+            self._spectrogram_to_mel_matrix(
+                num_spectrogram_bins=spectrogram.shape[1],
+                audio_sample_rate=audio_sample_rate,
+                **kwargs,
+            ).to(spectrogram),
+        )
+        return torch.log(mel_spectrogram + log_offset)
+
+    def _waveform_to_examples(self, data, sample_rate):
+        # Input: (T,) or (C, T)
+        # Convert to mono.
+        if len(data.shape) > 1:
+            data = torch.mean(data, axis=0)  # (T,)
```

Comment:
Stereo to mono conversion must be done carefully and taking mean is anti-pattern. See https://github.com/pytorch/audio/issues/877 for some discussion. We should land https://github.com/pytorch/audio/pull/3242, yet I am not sure if it is usable.

I suggest to remove stereo to mono conversion here and require user codes to convert their waveforms before passing them to the model.